openedx/core/djangoapps/discussions/docs/decisions/0004-in-context-discussions-linking.rst

```diff
@@ -0,0 +1,131 @@
+
+Currently this will only be supported for the edX Forums internal
+discussion provider, but it could be extended in the future to support
+other providers.
+
+In addition to providing a way to mark Units as discussable, there will
+also be a way to simply enable discussion for all units in the course.
+One further control provided here will be to toggle discussions on or
+off for graded discussions.
+
+Finally, it will also be possible to group discussions at a Sequence
+level instead of the Unit level.
+
+
+Requirements
+------------
+
+In order to enable in-context discussions with the above features we
+need a way to:
+
+- Mark a Unit as discussable
+- Store configuration options for discussions
+- Automatically mark/unmark units as discussable when settings change
+- Associate a unit with a corresponding discussion id
+- Make this discussion ID available via an API
+
+
+Consideration
+-------------
+
+Most of these configuration entries would be right at home in the
+`DiscussionsConfiguration` model in `plugin_settings`, however since they need
+to be available during course import-export, they should be stored in the
+course object itself.
+
+There is already a way to associate a discussion ID with a Discussion XBlock
+using its usage key. This same mechanism can be used to associate a Unit usage
+key with a corresponding discussion id.
+
+However the current mechanism has a few issues. It is stored as a JSON
+structure in the `DiscussionsIdMapping` model which has course id and a mapping
+of the discussion id to the xblock usage key in a single dict.
+
+This is OK for the existing setup because this is just a caching mechanism and
+the source of truth for this mapping is the XBlock itself, which stores the
+discussion id. On course publish this information is cached to
+`DiscussionsIdMapping`.
+
+For the new discussions system though, this mapping would be the source of
+truth for the link between discussions and units, so we should use a model
+where each such link is encoded as a row in the database.
+
+Decision
+--------
+
+Since the discussions settings need to be stored in the course structure we
+should create a new JSON structure in the course that matches the structure
+of `plugin_settings`. This can then be used to store not just the settings
+for the inbuilt discussions provider, but for any discussions provider in the
+future.
+
+When a course is published, we can copy over all the `plugin_settings` to the
+course structure with the following structure:
+
+.. code-block:: JSON
+
+    {
+        "edx-next": {
+            "discussions_enable_in_context": bool,
+            "discussions_enable_graded_units": bool,
+            "discussions_group_at_subsection": bool,
+            "discussions_custom_visibility": bool,
+        }
+    }
+
+The `edx-next` key here represents the provider id, allowing for potentially
+multiple provider configs to coexist in case of switching providers etc.
```

Comment:
Is there going to be any set of commonly respected configurations, or are all configurations going to be provider-dependent?

If there's some commonly respected subset, it might be worth doing something like:

```JavaScript
{
  "discussions": {
    "enable_in_context": true,
    "enable_graded_units": true,
    "group_at_subsection": true,
    "custom_visibility": true, 

    "providers": {
      "edx": {
        // cohorts, teams, etc.
      }
    }
  }
}
```

Reply:
Some of these could technically exist for other providers, but I don't think it's something that can easily carry over. i.e. in-context discussions will work quite differently between the inbuilt forums vs an LTI-based forum. I guess we can still keep these outside of provider-specific options, and they can be interpreted differently based on the current provider if it supports the option.

Reply:
I've rolled these up where they make sense.